includes/validation/class-amp-validation-manager.php

```diff
@@ -1585,11 +1586,53 @@ public static function get_source( $callback ) {
 
 		$source = compact( 'reflection' );
 
+
+		$file   = wp_normalize_path( $reflection->getFileName() );
+		$source = array_merge( $source, self::get_file_source( $file ) );
+
+		// If a file was identified, then also supply the line number.
+		if ( isset( $source['file'] ) ) {
+			$source['line'] = $reflection->getStartLine();
+		}
+
+		if ( $reflection instanceof ReflectionMethod ) {
+			$source['function'] = $reflection->getDeclaringClass()->getName() . '::' . $reflection->getName();
+		} else {
+			$source['function'] = $reflection->getName();
+		}
+
+		return $source;
+	}
+
+	/**
+	 * Identify the type, name, and relative path for a file.
+	 *
+	 * @since 2.0.1
+	 *
+	 * @param string $file File.
+	 * @return array {
+	 *     @type string $type Source type (core, plugin, mu-plugin, or theme).
+	 *     @type string $name Source name.
+	 *     @type string $file Relative file path based on the type.
+	 * }
+	 */
+	public static function get_file_source( $file ) {
+		if (
+			! isset( self::$template_directory )
+			||
+			! isset( self::$template_slug )
+			||
+			! isset( self::$stylesheet_directory )
+			||
+			! isset( self::$stylesheet_slug )
+		) {
+			self::set_theme_variables();
+		}
```

Comment:
I don't like this.

Reply:
What about moving the checks within the `set_theme_variables()` method?

Reply:
The ugly part here is that `set_theme_variables() `is an initialization method (like a constructor) that needs to be called for each execution of `get_file_source()` because the class is static and doesn't have any instantiation with a constructor.

To get around this, you could turn these static members into static methods with a caching static variable instead:
So, instead of referring to `self::$stylesheet_directory` in other parts of the code, you'd refer to `self::get_stylesheet_directory()` instead. Within these static methods, you can then do the initialization once and store it in a static variable.

However, I think that you're just trading multiple checks to a function call, so it's probably not a performance gain, only cleaner code.

Refactoring the entire validation manager into an instanced object (or a series of them) would be the solution that also improves performance.

Reply:
Something else that's tricky is that we need to wait until as late as possible to determine these values.

During validation requests, note how `set_theme_variables` gets called twice overall:

https://github.com/ampproject/amp-wp/blob/4ee9ee4ecc48431a26690245123b73f4b7154d1f/includes/validation/class-amp-validation-manager.php#L566-L570

This ensures that the theme is accurate when it is dynamically switched by plugins.

And calling it again when emitting the warning here will make sure that the theme variables get set to reflect the current state at the time the error happened.

Reply:
As it seems to make sense to reuse this in `ErrorPage`, I'd suggest we extract this out into a helper object instead.